### ISSUE TITLE
for...of array optimization can introduce conflicting variables into the scope

### ISSUE THREAD

**Issue by pomek** [babel/babel#8913]
## Bug Report

Initially, I reported this bug here – https://github.com/facebook/create-react-app/issues/5387 – because I thought that something is broken in CRA. After further investigation, I found a bug in the Babel so here I am.

**Current Behavior**

The following error.

```js
index.js:5 TypeError: Cannot read property '0' of undefined
    at Object.to (index.js:5)
    at new t (index.js:5)
    at new t (index.js:5)
    at new t (index.js:5)
    at e.t (index.js:5)
    at e (index.js:5)
    at new e (index.js:5)
    at index.js:5
    at new Promise (<anonymous>)
    at Function.value (index.js:5)
(anonymous) @ index.js:5
Promise.catch (async)
(anonymous) @ index.js:5
n @ index.js:1
(anonymous) @ index.js:5
n @ index.js:1
(anonymous) @ index.js:1
(anonymous) @ index.js:1
```

**Input Code**

I prepared a repository that allows reproducing the issue: https://github.com/pomek/react-app/tree/babel

_`#master` branch is reserved for CRA, this bug was committed to the `#babel` branch._

But in order to speed up resolving the issue, I'll describe everything in the issue as well.

Here is my `./src/index.js`:

```js
import ClassicEditor from '@ckeditor/ckeditor5-build-classic';

ClassicEditor.create( document.getElementById( 'editor' ) )
	.then( editor => {
		console.log( 'Ready to use!', editor );
	} )
	.catch( err => {
		console.error( err );
	} );

```

and `./dist/index.html`:

```html
<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="utf-8">
    <title>CKEditor 5 – classic editor build – development sample</title>
    <style>
        body {
            max-width: 800px;
            margin: 20px auto;
        }
    </style>
</head>
<body>

<h1>CKEditor 5 – classic editor build – development sample</h1>

<div id="editor">
    <h2>Sample</h2>

    <p>This is an instance of the <a href="https://ckeditor.com/docs/ckeditor5/latest/builds/guides/overview.html#classic-editor">classic editor build</a>.</p>

    <p>You can use this sample to validate whether your <a href="https://ckeditor.com/docs/ckeditor5/latest/builds/guides/development/custom-builds.html">custom build</a> works fine.</p>
</div>

<script src="./index.js"></script>

</body>
</html>
```

**Expected behavior/code**

An application should work:

![The working editor screen.](https://content.screencast.com/users/pomekPL/folders/Jing/media/4c7c15f7-390e-4ec4-ba82-a48959d9ec7f/00000059.png)

**Babel Configuration (.babelrc, package.json, cli command)**

I'm using Webpack to build the app:

`./webpack.config.js`

```js
module.exports = {
	context: __dirname,
	entry: [
		'@babel/polyfill',
		'./src/index.js'
	],
	output: {
		filename: './index.js'
	},

	module: {
		rules: [
			{
				test: /\.js/,
				use: {
					loader: 'babel-loader',
					options: {
						presets: [
							[ '@babel/preset-env', {
								targets: {
									ie: 9,
								},
								ignoreBrowserslistConfig: true,
								useBuiltIns: false,
								modules: false,
								exclude: [ 'transform-typeof-symbol' ],
							} ]
						]
					}
				}
			}
		]
	}
};
```

Webpack's config looks similar to config used by CRA: https://github.com/facebook/create-react-app/blob/7ba343c187744d077b49cfd3b68b805e497bf3ac/packages/babel-preset-react-app/dependencies.js#L83-L103

`./package.json`

```json
{
  "name": "react-app",
  "version": "1.0.0",
  "description": "",
  "dependencies": {
    "@ckeditor/ckeditor5-build-classic": "^11.1.1"
  },
  "devDependencies": {
    "@babel/cli": "^7.1.2",
    "@babel/core": "^7.1.2",
    "@babel/polyfill": "^7.0.0",
    "@babel/preset-env": "^7.1.0",
    "@babel/runtime": "^7.1.2",
    "babel-loader": "^8.0.4",
    "webpack": "^4.20.2",
    "webpack-cli": "^3.1.2"
  },
  "scripts": {
    "build": "webpack --mode=production"
  }
}
```

After installing dependencies, calling `npm run build` and opening `./dist/index.html` you should have a working editor. Unfortunately, you'll end with the error.

**Environment**
- Babel version(s): 7.1.2 (?)
- Node/npm version: 10.12.0 / 6.4.1
- OS: `20:13:40Darwin MacBook-Pro.local 18.0.0 Darwin Kernel Version 18.0.0: Wed Aug 22 20:13:40 PDT 2018; root:xnu-4903.201.2~1/RELEASE_X86_64 x86_64`
- How you are using Babel: `loader`

**Possible Solution**

Is no a solution but I wanted to check whether it will work if `ckeditor5-build-classic` won't be minified so I did steps specified below:

1. Go to `@ckeditor` namespace in dependencies: `cd node_modules/@ckeditor`
2. Remove `ckeditor5-build-classic` installed from npm: `rm -r ckeditor5-build-classic`
3. Clone package from GitHub and check out to the proper version: `git clone -b v11.1.0 git@github.com:ckeditor/ckeditor5-build-classic.git`
4. Go to the package and install dependencies: `cd ckeditor5-build-classic && npm i`
5. Build the package in development mode: `./node_modules/.bin/webpack --mode development`
6. Go back to your application and call `npm run build` again.

It works!

**Comment by loganfsmyth** [comment]
Yep, definitely looks like a Babel bug. Here's a reduced snippet of code that demonstrates the issue:
```
function delegate(...t){
  for(const o of t){
    const t = null;
  }
}
```
becomes
```
function delegate() {
  for (var _len = arguments.length, t = new Array(_len), _key = 0; _key < _len; _key++) {
    t[_key] = arguments[_key];
  }

  for (var _i = 0; _i < t.length; _i++) {
    var o = _t[_i];
    var _t = null;
  }
}
```
Note that the inner `t` was renamed to `_t` by the module transform, but it looks like `var o = t[_i];` was _also_ transformed when it should have been left referencing `t`.

**Comment by loganfsmyth** [comment]
Alright, what I'm thinking is that we should just avoid this whole issue if we can. It seems like the simplest way to do that would be to take
```
function delegate(...t){
  for(const o of t){
    const t = null;
  }
}
```
and instead of 
```
function delegate(...t) {
  for (var _i = 0; _i < t.length; _i++) {
    const o = t[_i];
    const t = null;
  }
}
```
we can instead output
```
function delegate(...t) {
  for (var _i = 0, o = t[_i]; _i < t.length; _i++, o = t[_i]) {
    const t = null;
  }
}
```
then we avoid needing any custom logic to handle scope-related things.

This should involve the logic in https://github.com/babel/babel/blob/9d0dcedb2b42c95e51a7892b5139b752a98f6e20/packages/babel-plugin-transform-for-of/src/index.js#L134,L159

I'm going to mark this as `good first issue` because it is selfcontained, but it will definitely require getting to understand our AST and how to manipulate it.

**Comment by kyle1320** [comment]
I'm going to take a shot at this.

**Comment by kyle1320** [comment]
@loganfsmyth I'm hesitant to use your proposed solution, because it would mean that code which relies on the scoping of the iteration variable could break if it gets replaced with `var`, e.g:
```javascript
var x = 'data';

for (let x of arr) {
  // do something with x
}

console.log(x); // should print 'data'
```
`x` would end up with the wrong value after the loop if the iterator was replaced with a `var` declaration.

And even if `var` is used, the following subtle bug would also be introduced:
```javascript
for (var x of arr) {
  // do something with x
}

console.log('last element is ' + x);
```
With the proposed solution, `x` would be `undefined` after the loop, instead of being the last element in the array.

I think the best solution would be to fix whatever scoping issue leads to the iterator declaration using the renamed variable, but I'm afraid that would be over my head for a first issue.

**Comment by loganfsmyth** [comment]
Hmm, that's a good point that I hadn't considered. That leaves us 2 main options.

1. Rename all uses of `t` inside the body of the loop to avoid the conflict
2. Detect the conflict and insert an extra scope, e.g.
    ```
    for (var _i = 0; _i < arr.len; _i++) {
      const t = arr[_i];
      {
        const t = null;
      }
    }
    ```

As long as we only add the extra scope when there are variable name conflicts, it won't affect 99% of people anyway, so I don't mind going with option 2.

Main thing I'll not on this is that we should make sure to use `path.getBindingIdentifiers()` when checking for conflicting variables, since the look could have multiple variables if it's `for (const {a, b} of arr) {}` or something.

**Comment by danez** [comment]
This seems it would be fixed by #9697
